client/mainmenu/CHighScoreScreen.cpp

```diff
@@ -0,0 +1,399 @@
+		{
+			std::string tmp = curData["land"].String();
+			TextOperations::trimRightUnicode(tmp, std::max(0, (int)TextOperations::getUnicodeCharactersCount(tmp) - 25));
+			texts.push_back(std::make_shared<CLabel>(405, y + i * 50, FONT_MEDIUM, ETextAlignment::CENTER, color, tmp));
+			texts.push_back(std::make_shared<CLabel>(557, y + i * 50, FONT_MEDIUM, ETextAlignment::CENTER, color, std::to_string(curData["days"].Integer())));
+			texts.push_back(std::make_shared<CLabel>(627, y + i * 50, FONT_MEDIUM, ETextAlignment::CENTER, color, std::to_string(curData["points"].Integer())));
+		}
+		else
+		{
+			std::string tmp = curData["campaign"].String();
+			TextOperations::trimRightUnicode(tmp, std::max(0, (int)TextOperations::getUnicodeCharactersCount(tmp) - 25));
+			texts.push_back(std::make_shared<CLabel>(405, y + i * 50, FONT_MEDIUM, ETextAlignment::CENTER, color, tmp));
+			texts.push_back(std::make_shared<CLabel>(592, y + i * 50, FONT_MEDIUM, ETextAlignment::CENTER, color, std::to_string(curData["points"].Integer())));
+		}
+
+		if(curData["points"].Integer() > 0 && curData["points"].Integer() <= ((highscorepage == HighScorePage::CAMPAIGN) ? 2500 : 500))
+			images.push_back(std::make_shared<CAnimImage>(AnimationPath::builtin("CPRSMALL"), (*CGI->creh)[HighScoreCalculation::getCreatureForPoints(curData["points"].Integer(), highscorepage == HighScorePage::CAMPAIGN)]->getIconIndex(), 0, 670, y - 15 + i * 50));
+	}
+}
+
+void CHighScoreScreen::buttonCampaignClick()
+{
+	highscorepage = HighScorePage::CAMPAIGN;
+	addHighScores();
+	addButtons();
+	redraw();
+}
+
+void CHighScoreScreen::buttonScenarioClick()
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL_NO_DISPOSE;
+	highscorepage = HighScorePage::SCENARIO;
+	addHighScores();
+	addButtons();
+	redraw();
+}
+
+void CHighScoreScreen::buttonResetClick()
+{
+	CInfoWindow::showYesNoDialog(
+		CGI->generaltexth->allTexts[666],
+		{},
+		[this]()
+		{
+			Settings entry = persistentStorage.write["highscore"];
+			entry->clear();
+			addHighScores();
+			addButtons();
+			redraw();
+		},
+		0
+	);
+}
+
+void CHighScoreScreen::buttonExitClick()
+{
+	close();
+}
+
+CHighScoreInputScreen::CHighScoreInputScreen(bool won, HighScoreCalculation calc)
+	: CWindowObject(BORDERED), won(won), calc(calc)
+{
+	addUsedEvents(LCLICK | KEYBOARD);
+
+	OBJ_CONSTRUCTION_CAPTURING_ALL_NO_DISPOSE;
+	pos = center(Rect(0, 0, 800, 600));
+	updateShadow();
+
+	background = std::make_shared<TransparentFilledRectangle>(Rect(0, 0, pos.w, pos.h), Colors::BLACK);
+
+	if(won)
+	{
+		int border = 100;
+		int textareaW = ((pos.w - 2 * border) / 4);
+		std::vector<std::string> t = { "438", "439", "440", "441", "676" }; // time, score, difficulty, final score, rank
+		for (int i = 0; i < 5; i++)
+			texts.push_back(std::make_shared<CMultiLineLabel>(Rect(textareaW * i + border - (textareaW / 2), 450, textareaW, 100), FONT_HIGH_SCORE, ETextAlignment::TOPCENTER, Colors::WHITE, CGI->generaltexth->translate("core.genrltxt." + t[i])));
```

Comment:
why MultiLineLabel?

Reply:
Original strings has a line break in it which has to be supported.